Simplified CSP algorithm

cle1109 opened the pull request:
Addresses #2437

agramfort commented on mne/decoding/csp.py:
```diff
@@ -115,33 +115,16 @@ def _fit(self, cov_a, cov_b):
+        w = w[:, ind]  # first, last, second, second last, third, ...
         self.filters_ = w
-        self.patterns_ = linalg.pinv(w).T
+        self.patterns_ = np.linalg.pinv(w)
```
we never use linalg from numpy. Please use the scipy module. Thx

agramfort commented:
do you changes make any difference numerically?

cle1109 commented:
Yes, numerically the new method produces patterns that are scaled versions of the patterns in the previous implementation. Qualitatively, scaling does not influence the eigenvectors so it doesn't matter.

agramfort commented:
examples render the same with same CV scores? just to be confident when
clicking the green button...

jasmainak commented:
oops, looks like I'm in sync with what @agramfort is thinking. Just posted at the wrong place :)

cle1109 commented:
Don't press that button just yet :-) - I'm running the examples right now...

cle1109 commented:
OK, after transposing the filters the classification results (including the time course) are now identical in the two CSP examples. The patterns are also identical in shape, but because the scaling is different, they look a bit different (I think the scaling in topomap is the same in both cases and isn't derived automatically from the min/max in the data).
In the attached pictures, the current results are in the upper (or left) panel and the new results in the lower (or right) panel.
![plot_decoding_csp_eeg_acc](https://cloud.githubusercontent.com/assets/4377312/11296440/7500fc7c-8f72-11e5-9239-f26b888532c2.png)
![plot_decoding_csp_eeg](https://cloud.githubusercontent.com/assets/4377312/11296441/750236c8-8f72-11e5-8b6a-9bee0e214611.png)
![plot_decoding_csp_space](https://cloud.githubusercontent.com/assets/4377312/11296442/75063b10-8f72-11e5-8ecd-e4b7c4dbc0dd.png)

agramfort commented:
hum so it's changing slightly the results...

so when you say simplify do you say speed up?

I just want to avoid changing code behavior for cosmetic reasons.

cle1109 commented:
By simplify I mean readability. Basically, I can replace several steps by using the generalized eigenvalue problem in one line. However, I'm not sure why the scaling is different. Maybe it's possible to obtain the same results if we scale the eigenvectors differently, I will try but can't promise anything.

@BenjaminBlankertz, what do you think? Should we keep using the current implementation or switch to the shorter generalized eigenvalue solution?

agramfort commented:
@BenjaminBlankertz <https://github.com/BenjaminBlankertz> it would be great
to have your input here.

mbillingr commented:
I wonder why the previous implementation was more complicated.

Obviously, the author was aware of the `eigh`'s ability to perform the generalized eigenvalue decomposition (line 131). So, why the two-stepped approach?

cle1109 commented:
I think because that's how Koles described it in his 1990 paper (although IIRC Fukunaga, who is usually cited in the context of simultaneous diagonalization for CSP, mentions the equivalence to the generalized eigenvalue problem in his 1972 book).

mbillingr commented:
How does Koles motivate the approach?

I fear there may be a good reason for it, like numerical stability or something.

cle1109 commented:
He doesn't, he just follows the approach for simultaneous diagonalization. Maybe he didn't read the "Alternative approach" section in Fukunaga's book chapter but just implemented the first version :-P. I think both approaches are really identical except one is a shorthand version of the other. I don't think that there are numerical stability issues, but since there are less eigenvalue decompositions involved in the generalized approach, there should be less numerical problems with this implementation than with the multi-step one.

BenjaminBlankertz commented:
I am clearly in favor of the simple implementation. (We don't even do the normalization by trace.)
Without a deeper look, I cannot comment on the differences in scaling and sign, but I strongly assume that they are not relevant.

Btw: I think the resorting
  w = w[:, ::-1]  # descending order
could be saved by doing that implicitly in the selection step
  w = w[:, ind]
when indices 'ind' are chosen accordingly. Don't know whether that would be a relevant saving for mne.

Eric89GXL commented:
I can't add much to this discussion, other than to mention that `w = w[:, ::-1]` won't copy the data but just make a view, so it should be very fast. The only practical disadvantage I know of doing it is that it can make some subsequent operations (e.g., reshape) force a copy, and can prevent SIMD instructions from MKL libraries from being used, but I doubt either of those apply too much here.

agramfort commented:
thanks @BenjaminBlankertz for pitching in

so we're good here? shall we update the reference in the docstring to better point towards the paper we are actually implementing here?

cle1109 commented:
Thanks @BenjaminBlankertz, I added your tutorial paper as a reference. I also replaced the 1991 Koles paper with his 1990 paper, because the latter was really the first one to introduce CSP analysis to EEG data.

BenjaminBlankertz commented:
I understand that omitting `w = w[:, ::-1]` doesn't bring a relevant gain. However, for the sake of clean code: it is not clear to me, why this is done anyway. Isn't this reversing just the same as exchange the roles of the two classes (since their Eigenvalues add up to 1)? If I am right, then it we could write `eigh(cov_b, cov_a + cov_b)` instead of `eigh(cov_a, cov_a + cov_b)` and omit the reversing. But maybe I am missing something.

mbillingr commented:
I think @BenjaminBlankertz is correct.

Later the eigen vectors are rearranged again with `w = w[:, ind]  # first, last, second, second last, third, ...`
If the order of classes is important the index array could be changed to `last, first, second last, second, ...` and removing the `w = w[:, ::-1]` statement.

agramfort commented:
agreed. travis is is not happy but it's maybe unrelated.

cle1109 commented:
Yes, I agree with @BenjaminBlankertz and @mbillingr - should we swap `cov_a` and `cov_b` in the eigenvalue computation? Or adapt the `ind` array?

cle1109 commented:
Regarding the AppVeyor error: it looks like Python 2.7 doesn't support UTF8 source files. How is this handled in MNE?

agramfort commented:
don't use utf8 then :)

cle1109 commented:
So in other words, ü = ue?

Eric89GXL commented:
You can use utf8 but you need an appropriate line at the top of the file.
See preprocessing/maxwell.py

agramfort commented:
>
> So in other words, ü = ue?
>
yes :)

agramfort commented:
you have a travis failure due to flake8 violation.

cle1109 commented:
Wow, I really need to ramp up my test suite before I push my commits... Anyway, everything seems to be working now.

agramfort commented:
looks good.

thanks @cle1109 !

agramfort commented:
crap now in master we have pb... plot_decoding_csp_eeg.py broken.

ValueError: could not broadcast input array from shape (31) into shape (32)

can you have a look ASAP?

and confirm that all examples run fine?

thx

cle1109 commented:
Sorry about that - reversing the indexing wasn't that straightforward. I fixed it, the examples run fine, and pushed it, but I don't know how to fix this now.

agramfort commented:
Open a new pr 

Even maybe add a test so this does not happen in the future 

> On 24 nov. 2015, at 22:56, Clemens Brunner <notifications@github.com> wrote:
> 
> Sorry about that - reversing the indexing wasn't that straightforward. I fixed it, the examples run fine, and pushed it, but I don't know how to fix this now.
> 
> —
> Reply to this email directly or view it on GitHub.
>
